sky/provision/gcp/instance.py

```diff
@@ -72,6 +90,249 @@ def _wait_for_operations(
+                cluster_name_on_cloud,
+                project_id,
+                availability_zone,
+                resumed_instance_ids[0],
+                is_head=True,
+            )
+
+    if to_start_count > 0:
+        results = resource.create_instances(cluster_name_on_cloud, project_id,
+                                            availability_zone,
+                                            config.node_config, labels,
+                                            to_start_count,
+                                            head_instance_id is None)
+        if head_instance_id is None:
+            head_instance_id = results[0][1]
+        # FIXME: it seems that success could be False sometimes, but the instances
+        #   are started correctly.
+        created_instance_ids = [instance_id for success, instance_id in results]
+
+    while True:
+        # wait until all instances are running
+        instances = resource.filter(
+            project_id=project_id,
+            zone=availability_zone,
+            label_filters=filter_labels,
+            status_filters=PENDING_STATUS,
+        )
+        if not instances:
+            break
+
+    return common.ProvisionRecord(provider_name='gcp',
+                                  region=region,
+                                  zone=availability_zone,
+                                  cluster_name=cluster_name_on_cloud,
+                                  head_instance_id=head_instance_id,
+                                  resumed_instance_ids=resumed_instance_ids,
+                                  created_instance_ids=created_instance_ids)
+
+
+def wait_instances(region: str, cluster_name_on_cloud: str,
+                   state: Optional[status_lib.ClusterStatus]) -> None:
+    """See sky/provision/__init__.py"""
+    # We already wait for the instances to be running in run_instances.
+    # So we don't need to wait here.
+    return
+
+
+def get_cluster_info(
+        region: str,
+        cluster_name_on_cloud: str,
+        provider_config: Optional[Dict[str, Any]] = None) -> common.ClusterInfo:
+    """See sky/provision/__init__.py"""
+    assert provider_config is not None, cluster_name_on_cloud
+    zone = provider_config['availability_zone']
+    project_id = provider_config['project_id']
+    label_filters = {TAG_RAY_CLUSTER_NAME: cluster_name_on_cloud}
+
+    handlers: List[Type[instance_utils.GCPInstance]] = [
+        instance_utils.GCPComputeInstance
+    ]
+    use_tpu_vms = provider_config.get('_has_tpus', False)
+    if use_tpu_vms:
+        handlers.append(instance_utils.GCPTPUVMInstance)
+
+    handler_to_instances = _filter_instances(
+        handlers,
+        project_id,
+        zone,
+        label_filters,
+        lambda _: ['RUNNING'],
+    )
+    instances = {}
+    for res, insts in handler_to_instances.items():
+        for inst in insts:
+            instances[inst] = res.get_instance_info(project_id, zone, inst)
```

Comment:
This part seems taking quite a long time, do we want to parallelize it a bit?
```suggestion
        with pool.ThreadPool() as p:
            instances.update(
                dict(zip(insts, p.starmap(res.get_instance_info,
                                       [(project_id, zone, inst) for inst in insts]))))
```